### DISCUSSION TITLE
Validation error messages use the field name instead of the field label

### DISCUSSION THREAD

**Discussion by bemdesign-er** [pods-framework/pods#6685] [Ideas]
**Is your feature request related to a problem? Please describe.**
I would like the ability to change or modify field validation messages on a per-field basis. This is particularly useful on select fields where normally the validation message uses the not-always-user-friendly programatic name for the field.

**Describe the solution you'd like**
When setting up a field for a Pods item, have an additional area to capture pod-editor specified error validation message for that field.

**Describe alternatives you've considered**
Custom Javascript after the validation error message is generated to change the message - cons with this approach is that it's not as user friendly to do (requires javascript programming) and can be problematic with timing (identifying when the error validation message is shown and when to fire off the error message customization script).

**Additional context**
The use of progromatic field names for validation error messages may be a limitation of the js validation library being used.

**Comment by pglewis** [comment]
All fields will be getting a make-over as part of 2.8 (we have an issue for every field on the 2.8 milestone, eg #4662).  That's a perfect time to address this.

**Comment by sc0ttkclark** [comment]
@pdclark Can you check and see if we are using field labels where they should be used now for required field error messages?

@bemdesign-er for custom error messages on submit, you can try this:

```
add_filter( 'pods_field_{type}_validate', static function( $validate, $value, $name, $options, $fields, $pod, $id, $type, $params ) {
    if ( false === $validate || is_array( $validate ) ) {
        return $validate;
    }

    if ( 'my_field' === $name && '123456' === $value ) {
        $errors = [];

        $errors[] = 'You cannot enter "123456" for this field.';

        return $errors;
    }
}, 10, 9 );
```

**Comment by pdclark** [comment]
- Labels appear in the error messages.
- Filtering the error message doesn't appear to work, or rather, the custom error message is logged to PHP error log, but not output to UI (see screenshot).
- Relationship fields don't appear to do anything in response to required. Or rather, single-select doesn't provide an option to select none, and multi-select doesn't complain if none are selected.

`'pods_field_{type}_validate'` with 9 params didn't appear to do anything for me.
Instead I found `'pods_field_validate_{type}'` with 8 params:

e.g.,
```php
<?php

add_filter( 'pods_field_validate_text', static function( $validate, $value, $name, $options, $fields, $pod, $id, $params ) {
	if ( false === $validate || is_array( $validate ) ) {
		return $validate;
	}

	if ( 'required_plain_text' === $name && '123456' === $value ) {
		$errors = [];

		$errors[] = 'You cannot enter "123456" for this field.';

		return $errors;
	}
}, 10, 8 );
```

Which kind of worked — but the error only went to PHP error log, not to UI.

<img width="1679" alt="Screen Shot 2021-09-27 at 11 25 01 AM" src="https://user-images.githubusercontent.com/126889/134964937-67995cc7-d4d2-4659-833e-eaa9b81714a5.png">

@sc0ttkclark

**Comment by JoryHogeveen** [comment]
@pdclark @sc0ttkclark 
Indeed, it should be `pods_field_validate_{$type}`: https://github.com/pods-framework/pods/blob/main/classes/PodsField.php#L606-L620

However, I also found a filter in PodsForm: https://github.com/pods-framework/pods/blob/main/classes/PodsForm.php#L1180

@sc0ttkclark We'll need to address this as there are more filters in PodsForm that might relate to Field filters..

EDIT: Added a separate issue to address this.

**Comment by JoryHogeveen** [comment]
As for the error log in the UI.
Honestly, if I'm testing locally on 2.8 error reporting doesn't even work at all on the block editor @sc0ttkclark 
On the classic editor I get a WP Error page (which is very inconvenient since all data is lost)

Related: #5464  (@sc0ttkclark why is this punted? IMO this is a blocking issue)

**Comment by sc0ttkclark** [comment]
@JoryHogeveen Both of these issues seem to have existed before Pods 2.8.

There's more work to be done on the React side of things for all forms to make a REST API validation call before save to be able to prevent the save itself from occurring on any form if it fails initial server validation. Once that is done, we can look at addressing this and #5464 more fully.
